crates/bevy_ecs/src/component.rs

```diff
@@ -1133,3 +1145,131 @@ impl<T: Component> FromWorld for InitComponentId<T> {
         }
     }
 }
+#[cfg(feature = "track_change_detection")]
+type RequiredComponentConstructor =
+    dyn Fn(&mut Table, &mut SparseSets, Tick, TableRow, Entity, &'static Location<'static>);
+#[cfg(not(feature = "track_change_detection"))]
+type RequiredComponentConstructor = dyn Fn(&mut Table, &mut SparseSets, Tick, TableRow, Entity);
+
+/// A context-less constructor that can be called to construct a new instance of a component.
+#[derive(Clone)]
+pub(crate) struct RequiredComponent {
+    pub(crate) component_id: ComponentId,
+    /// # Safety
+    /// Calling this constructor is unsafe. It should only be called in the context of [`BundleInfo::write_components`], where the
+    /// inputs are already validated. This _should not_ have its module visibility increased.
+    constructor: Arc<RequiredComponentConstructor>,
+}
+
+impl RequiredComponent {
+    /// # Safety
+    /// This is intended to only be called in the context of [`BundleInfo::write_components`] to initialized required components.
+    /// Calling it _anywhere else_ should be considered unsafe.
+    ///
+    /// `table_row` and `entity` must correspond to a valid entity that currently needs a component initialized via the constructor stored
+    /// on this [`RequiredComponent`]. The stored [`RequiredComponent::constructor`] must correspond to a component on `entity` that needs initialization.
+    /// `table` and `sparse_sets` must correspond to storages on a world where `entity` needs this required component initialized.
+    ///
+    /// Again, don't call this anywhere but [`BundleInfo::write_components`].
+    pub(crate) unsafe fn initialize(
+        &self,
+        table: &mut Table,
+        sparse_sets: &mut SparseSets,
+        change_tick: Tick,
+        table_row: TableRow,
+        entity: Entity,
+        #[cfg(feature = "track_change_detection")] caller: &'static Location<'static>,
+    ) {
+        (self.constructor)(
+            table,
+            sparse_sets,
+            change_tick,
+            table_row,
+            entity,
+            #[cfg(feature = "track_change_detection")]
+            caller,
+        );
+    }
+}
+
+/// The collection of metadata for components that are required for a given component.
```

Comment:
This either needs a doc test demonstrating how to use the macro to register required components or a link to the docs where this *is* demonstrated.